lib/EmitVerilog/EmitVerilog.cpp

```diff
@@ -1922,93 +1971,101 @@ void ModuleEmitter::collectNamesEmitDecls(Block &block) {
+      // base name may be unique but the suffixed names may not be.  The right
+      // way to solve this is to change the instances and mems in a new
+      // Verilog dialect to use multiple return values, exposing the
+      // independent Value's.
 
-    // Determine what kind of thing this is, and how much space it needs.
-    maxDeclNameWidth =
-        std::max(getVerilogDeclWord(&op).size(), maxDeclNameWidth);
+      // Determine what kind of thing this is, and how much space it needs.
+      maxDeclNameWidth =
+          std::max(getVerilogDeclWord(&op).size(), maxDeclNameWidth);
 
-    // Flatten the type for processing of each individual element.
-    fieldTypes.clear();
-    flattenBundleTypes(result.getType(), "", false, fieldTypes);
+      // Flatten the type for processing of each individual element.
+      fieldTypes.clear();
+      flattenBundleTypes(result.getType(), "", false, fieldTypes);
 
-    // Handle the reg declaration for a memory specially.
-    if (auto memOp = dyn_cast<MemOp>(&op))
-      if (auto dataType = memOp.getDataTypeOrNull())
-        flattenBundleTypes(dataType, "", false, fieldTypes);
+      // Handle the reg declaration for a memory specially.
+      if (auto memOp = dyn_cast<MemOp>(&op))
+        if (auto dataType = memOp.getDataTypeOrNull())
+          flattenBundleTypes(dataType, "", false, fieldTypes);
 
-    for (const auto &elt : fieldTypes) {
-      int bitWidth = getBitWidthOrSentinel(elt.type);
-      if (bitWidth == -1) {
-        emitError(&op, getName(result))
-            << elt.suffix << " has an unsupported verilog type " << elt.type;
-        result = Value();
-        break;
-      }
+      for (const auto &elt : fieldTypes) {
+        int bitWidth = getBitWidthOrSentinel(elt.type);
+        if (bitWidth == -1) {
+          emitError(&op, getName(result))
+              << elt.suffix << " has an unsupported verilog type " << elt.type;
+          result = OpResult();
+          break;
+        }
 
-      if (bitWidth != 1) { // Width 1 is implicit.
-        // Add 5 to count the width of the "[:0] ".
-        size_t thisWidth = getPrintedIntWidth(bitWidth - 1) + 5;
-        maxTypeWidth = std::max(thisWidth, maxTypeWidth);
+        if (bitWidth != 1) { // Width 1 is implicit.
+          // Add 5 to count the width of the "[:0] ".
+          size_t thisWidth = getPrintedIntWidth(bitWidth - 1) + 5;
+          maxTypeWidth = std::max(thisWidth, maxTypeWidth);
+        }
       }
-    }
 
-    // Emit this declaration.
-    if (result)
-      declsToEmit.push_back(&op);
+      // Emit this declaration.
+      if (result)
+        declsToEmit.push_back(result);
+    }
   }
 
   SmallPtrSet<Operation *, 8> ops;
 
   // Okay, now that we have measured the things to emit, emit the things.
-  for (auto *decl : declsToEmit) {
+  for (auto result : declsToEmit) {
     ops.clear();
-    ops.insert(decl);
+    ops.insert(result.getDefiningOp());
```

Comment:
I don't think this is right in the face of multiple result instances, because it will process each instance op multiple times.

Reply:
`ops` is a set so I don't think so. Either way, some of the above my obviate this change.